File: pset6/sentimental/vigenere/vigenere.py
# Implement vigenere's algorithm

import sys
from cs50 import get_string

def main():
    # Check command-line arguments
    if len(sys.argv) != 2:
        print("Usage: python vigenere keyword")
        sys.exit(1)

    keyword = sys.argv[1]
    if not keyword.isalpha():
        print("The keyword must contain only alphabetic characters")
        sys.exit(2)

    plaintext = get_string("plaintext: ")

    ciphered = cipher_vigenere(plaintext, keyword)

    print(f"ciphered: {ciphered}")

# Eipher text using the Vigenere algorithm
def cipher_vigenere(text, key):
    # Get all shifts of key
    keyshifts = list(ord(x.lower()) - 97 for x in key)

    # Prepare to word
    keyword_len = len(keyshifts)
    ciphered = list(text)
    j = 0

    # cipher every character
    for i in range(len(ciphered)):
        char = ciphered[i]
        # Only for alphabetic characters
        if not char.isalpha:
            continue

        # End of keyword
        if j >= keyword_len:
            j = 0

        # Store offset of case
        offset = 65 if ciphered[i].isupper() else 97

        shift = (ord(char) - offset + keyshifts[j]) % 26
        ciphered[i] = chr(offset + shift)
        j += 1
    return ''.join(ciphered)

if __name__ == "__main__":
    main()

File: pset2/vigenere/vigenere.c
#include <cs50.h>
#include <stdio.h>
#include <string.h>
#include <ctype.h>

bool is_valid_args(int, string[]);
string encipher(string, string);
bool isalpha_string(string);

int main(int argc, string argv[])
{
    string input, enciphered;

    if (!is_valid_args(argc, argv))
    {
        printf("Invalid arguments.\n");
        return 1;
    }

    input = get_string("plaintext:");

    enciphered = encipher(input, argv[1]);

    // Print the result.
    printf("enciphered: %s\n", enciphered);
}

// Check key
bool is_valid_args(int argc, string argv[])
{
    return argc == 2 && isalpha_string(argv[1]);
}

// Check that the string contains only alphabetic symbols.
bool isalpha_string(string text)
{
    int i;

    for (i = 0; text[i] != '\0'; i++)
        if (!isalpha(text[i]))
            return false;
    return true;
}

// Encipher text with Vigenere algorithm.
string encipher(string text, string key)
{
    int i, j, shift;

    for (i = j = 0; text[i] != '\0'; i++)
    {
        // if non-alphabetic symbols
        if (!isalpha(text[i])) continue;

        if (key[j] == '\0') j = 0;
        shift = (toupper(text[i]) + toupper(key[j])) % 26;
        if (isupper(text[i]))
            text[i] = 'A' + shift;
        else
            text[i] = 'a' + shift;

        j++;
    }

    return text;
}